feat: Vault entity page (#199)
Kong/developer.konghq.com: app/_gateway_entities/vault.md

in declarative configuration files, logs, or in the UI.

For example, you could store a certificate and a key in a Vault, then reference them from a [Certificate entity](/gateway/entities/certificate/). This way, the certificate and key are not stored in the entity directly and are more secure.

## How do I add secrets to a Vault?

Vaults store secrets, they don't create secrets. You must add secrets to the Vault in one of the following ways: 
* Environment variables
* {{site.konnect_short_name}} Config Store
* Supported third-party backend vault

The following table goes into detail about the different options you have:

| Backend                                                   | {{site.base_gateway}} OSS | {{site.base_gateway}} Enterprise | Uses Vault entity              | {{site.konnect_short_name}} supported |
| --------------------------------------------------------- | ------------------------ | -------------------------------- | ------------------------------ | ------------------------------------- |

> | [Environment variable](/how-to/store-secrets-as-env-variables) | <img src="/app/assets/icons/check.svg" alt="Check icon">   | <img src="/app/assets/icons/check.svg" alt="Check icon">   | <img src="/app/assets/icons/close.svg" alt="Cross icon">   | <img src="/app/assets/icons/check.svg" alt="Check icon">        |

Author: The check marks aren't rendering, does anyone know what I did wrong?

Reviewer: I just pushed a commit to your branch, please double-check if the data is correct.
I added `{% feature_table %}` block (it works in the same way as the component we use on landing pages), it takes a yaml config and renders the table

```
{vault://hcv/pg/username}
```

Would point to a secret object called `pg` inside a HashiCorp Vault, which may return the following value:

```json
{
  "username": "john",
  "password": "doe"
}
```

<!-- vale off -->
{{site.base_gateway}} receives the payload and extracts the `"username"` value of `"john"` for the secret reference of
`{vault://hcv/pg/username}`.
<!-- vale on -->

## How do I manage secrets stored in a Vault?

Although {{site.base_gateway}} can't create or add secrets to a Vault, it can help you [manage your secrets](/secrets-management/). By default, {{site.base_gateway}} automatically rotates secrets *once every minute* in the background. You can also further configure how often {{site.base_gateway}} rotates secrets using the Vault entity configuration and if it rotates secrets on failure (such as a database authentication failure).

For more information, see [Secret rotation](/gateway/secrets-management/secret-rotation/).

## Schema

The Vault entity can only be used once the database is initialized. Secrets for values that are used before the database is initialized can’t make use of the Vaults entity.

> {% entity_schema %}

Author: The schema won't render, any idea why?

Reviewer: the schema does render for me both  locally and in the preview app. do you have `VITE_PORTAL_API_URL` set in an `.env` file?

---

> title: Vaults

Author: This will be in the generated schema once the source file is fixed (mentioned in this Slack thread here: https://kongstrong.slack.com/archives/G012HLJ4GJJ/p1736455142843339)

content_type: reference
entities:
  - vault

description: A Vault is used to store secrets.

related_resources:
  - text: Vault CLI

>     url: /

Reviewer 2: should we have an empty page for Vault CLI and link to it instead?

> related_resources:
>   - text: Vault CLI
>     url: /

Reviewer 3: ```suggestion
related_resources:
  - text: Vault CLI
    url: /
```
Add workspaces, rbac, secrets management stubs

  - text: Secrets Management
    url: /secrets-management
  - text: Secret rotation
    url: /gateway/secret-management/secret-rotation
  

faqs:

>   - q: I'm using declarative configuration for my Vault. Should I split my Vault configuration?

Reviewer 3: If it means splitting into a different vault configuration, we should have a section in this doc that explains that.

>   - q: I'm using declarative configuration for my Vault. Should I split my Vault configuration?

Author: I thought I understood, but going back over the original content, I don't think I understood what splitting was. I asked in the API Ops channel: https://kongstrong.slack.com/archives/C04349E4KRC/p1736457078457189

    a: |
        For larger teams with many contributors, or organizations with multiple teams, we recommend splitting Vault configuration and managing it separately. We recommend splitting the configuration for the following reasons:
        * Vault are closer to infrastructure than other {{site.base_gateway}} configurations.
        Separation of routing policies from infrastructure-specific configurations helps
        keep configuration organized.
        * Vaults may be shared across teams. In this case, one specific team shouldn't
        control the Vault's configuration. One team changing the Vault a can have
        disastrous impact on another team.
        * If a Vault is deleted while in use -- that is, if there are still references to
        secrets in a Vault in configuration -- it can lead to total loss of proxy capabilities.
        Those secrets would be unrecoverable.

>   - q: How should I manage my Vault configuration with decK?

Reviewer 3: ```suggestion
  - q: What are general best practices for managing Vaults
```

    a: |
        To keep your environment secure and avoid taking down your proxies by accident, make sure to:
        * Manage Vaults with distributed configuration via tags.
        * Use a separate [RBAC role, user, and token](/gateway/entities/rbac/)
        to manage Vaults. Don't use a generic admin user.

>         * Set up a separate CI pipeline for Vaults.

Reviewer 3: I found the source of this: 
![image](https://github.com/user-attachments/assets/f6a8ffe8-dd81-4ae7-af0e-ef1c124a16be)

I don't think this actually answers the Q here. This sounds more like "what are some best practices for managing Vaults"

    uses_vault: true
    supports_konnect: true
  - title: Hashicorp Vault
    url: /how-to/configure-hashicorp-vault-as-a-vault-backend/
    oss: false
    enterprise: true
    uses_vault: true
    supports_konnect: true
{% endfeature_table %}

## How do I use secrets stored in a Vault?

When you want to use a secret stored in a Vault, you can reference the secret with a `vault` reference. You can use the `vault` reference in places such as `kong.conf`, declarative configuration files, logs, or in the UI.

A secret reference points to a string value. No other data types are currently supported.

The Vault backend may store multiple related secrets inside an object, but the reference
should always point to a key that resolves to a string value. For example, the following reference:

```
{vault://hcv/pg/username}
```

Would point to a secret object called `pg` inside a HashiCorp Vault, which may return the following value:

```json
{
  "username": "john",
  "password": "doe"
}
```

<!-- vale off -->
{{site.base_gateway}} receives the payload and extracts the `"username"` value of `"john"` for the secret reference of
`{vault://hcv/pg/username}`.
<!-- vale on -->

> ## How do I manage secrets stored in a Vault?

Reviewer 3: ```suggestion
## Secret rotation in Vault
```

> Although {{site.base_gateway}} can't create or add secrets to a Vault, it can help you [manage your secrets](/secrets-management/). By default, {{site.base_gateway}} automatically rotates secrets *once every minute* in the background. You can also further configure how often {{site.base_gateway}} rotates secrets using the Vault entity configuration and if it rotates secrets on failure (such as a database authentication failure).

Reviewer 3: ```suggestion
By default, {{site.base_gateway}} automatically rotates secrets *once every minute* in the background. You can also configure how often {{site.base_gateway}} rotates secrets using the Vault entity configuration. 

There are two types of rotation configuration available: 
Rotate periodically using TTLs (e.g. check for new TLS certificate once per day)
Rotate on failure (e.g. on database authentication failure, check if the secrets were updated, and try again)
```

This paragraph is about secret rotation. Not about managing secrets stored in a vault.
I can't find any information about the vault entity configuration being used to rotate secrets, i checked kong.conf and the API. I think you can do this in the individual backends but I do'nt know. Vault config that changes the rotation should be in this doc.

Author: > I can't find any information about the vault entity configuration being used to rotate secrets, i checked kong.conf and the API. I think you can do this in the individual backends but I do'nt know.

@Guaris It's here: https://docs.konghq.com/gateway/latest/kong-enterprise/secrets-management/advanced-usage/#shared-configuration-parameters:~:text=Most%20of%20the%20vaults%20also%20support%20secret%20rotation%20by%20using%20TTLs%3A

The config in the API schema just provides a link to that doc instead of listing out the different config options and I don't know why.
<img width="1065" alt="Screenshot 2025-01-09 at 11 20 16 AM" src="https://github.com/user-attachments/assets/2f36e11c-9b18-4ca8-9ff7-4c54cd5c78e1" />

For more information, see [Secret rotation](/gateway/secrets-management/secret-rotation/).

## Schema

The Vault entity can only be used once the database is initialized. Secrets for values that are used before the database is initialized can’t make use of the Vaults entity.

{% entity_schema %}

> ## Set up a Vault

Author: It was an issue with the liquid variable there, Fabian helped me fix it so it should render now

> ## Set up a Vault

Author: Thanks for catching this! It worked fine for me with decK. I think it's missing `/vaults` in the url in the rendered entity example, but I also adjusted how the parameters are displayed. 

@fabianrbz Do you know why /vaults is missing from the Admin API URL example? 
<img width="1030" alt="Screenshot 2025-01-09 at 3 41 51 PM" src="https://github.com/user-attachments/assets/0613ac9e-c66c-4377-8174-1f4a24e8a233" />

> ## Set up a Vault

Reviewer: @Guaris @cloudjumpercat I just fixed that [here](https://github.com/Kong/developer.konghq.com/pull/199/commits/a4e656c086aec707693cccbe2d89cb2b8466b033), in case this comes up for other entity
